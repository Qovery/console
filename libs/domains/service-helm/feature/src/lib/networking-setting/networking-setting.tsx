import { type HelmPortRequestPortsInner } from 'qovery-typescript-axios'
import { type PropsWithChildren } from 'react'
import { SettingsHeading } from '@qovery/shared/console-shared'
import {
  BlockContent,
  Button,
  ExternalLink,
  Heading,
  Icon,
  Popover,
  Section,
  useModal,
  useModalConfirmation,
} from '@qovery/shared/ui'
import { NetworkingPortSettingModal } from '../networking-port-setting-modal/networking-port-setting-modal'

export interface NetworkingSettingProps extends PropsWithChildren {
  helmId: string
  ports: HelmPortRequestPortsInner[]
  onUpdatePorts: (ports: HelmPortRequestPortsInner[]) => void
  isSetting?: boolean
}

export function NetworkingSetting({
  helmId,
  ports,
  onUpdatePorts,
  isSetting = false,
  children,
}: NetworkingSettingProps) {
  const { openModal, closeModal } = useModal()
  const { openModalConfirmation } = useModalConfirmation()

  const onAddPort = () =>
    openModal({
      content: (
        <NetworkingPortSettingModal
          helmId={helmId}
          onSubmit={(port) => {
            onUpdatePorts([...ports, port])
            closeModal()
          }}
          onClose={closeModal}
        />
      ),
    })
  const onEditPort = (originalPort: HelmPortRequestPortsInner) =>
    openModal({
      content: (
        <NetworkingPortSettingModal
          helmId={helmId}
          port={originalPort}
          onSubmit={(port) => {
            onUpdatePorts([...ports.filter((p) => p !== originalPort), port])
            closeModal()
          }}
          onClose={closeModal}
        />
      ),
    })
  const onRemovePort = (port: HelmPortRequestPortsInner) =>
    openModalConfirmation({
      title: 'Delete Port',
      isDelete: true,
      action: () => {
        onUpdatePorts(ports.filter((p) => p !== port))
      },
    })

  return (
    <Section className="items-start">
      {isSetting ? (
        <SettingsHeading
          title="Networking"
          description="You can expose publicly over HTTP/gRPC the Kubernetes services deployed."
        >
          <Button className="gap-2" size="md" variant="solid" color="brand" onClick={onAddPort}>
            Add Port
            <Icon iconName="plus-circle" iconStyle="regular" />
          </Button>
        </SettingsHeading>
      ) : (
        <div className="flex w-full flex-row justify-between">
          <div>
            <Heading className="mb-2">Networking</Heading>
            <p className="mb-2 text-sm text-neutral-350">
              You can expose publicly over HTTP/gRPC the Kubernetes services deployed.
            </p>
            <Popover.Root>
              <Popover.Trigger>
                <span className="mb-5 cursor-pointer text-sm font-medium text-brand-500 transition hover:text-brand-600">
                  How it works <Icon className="text-xs" iconStyle="regular" iconName="circle-question" />
                </span>
              </Popover.Trigger>
              <Popover.Content side="left" className="relative text-sm text-neutral-350" style={{ width: 440 }}>
                <h6 className="mb-2 font-medium text-neutral-400">How it works</h6>
                <p>
                  You can expose publicly over HTTP/gRPC the Kubernetes services deployed with this helm chart by
                  providing the service name (as defined within the chart) and the port you want to expose. Qovery will
                  take care of assigning a domain and configure the TLS for you.
                </p>
                <ExternalLink href="https://hub.qovery.com/docs/using-qovery/configuration/helm/#ports">
                  Documentation
                </ExternalLink>
                <Popover.Close className="absolute right-4 top-4">
                  <button type="button">
                    <Icon iconName="xmark" className="text-lg leading-4 text-neutral-400" />
                  </button>
                </Popover.Close>
              </Popover.Content>
            </Popover.Root>
          </div>
          <Button className="gap-2" size="lg" variant="solid" color="brand" onClick={onAddPort}>
            Add port
            <Icon iconName="plus-circle" />
          </Button>
        </div>
      )}
      <form className="w-full">
        <BlockContent title="Services exposed publicly" classNameContent="p-0">
          {ports.length > 0 ? (
            ports.map((port) => {
              const { service_name, internal_port, protocol, namespace, name } = port
              return (
                <div
                  className="flex w-full items-center justify-between gap-3 border-b border-neutral-250 px-5 py-4 last:border-0"
                  key={name}
                >
                  <div className="flex flex-col">
                    <span className="text-sm font-medium text-neutral-400">{service_name}</span>
                    <div className="flex flex-row gap-2 text-xs text-neutral-350">
                      <span>Service port: {internal_port}</span>
                      <span>Protocol: {protocol}</span>
                      {namespace && <span>Namespace: {namespace}</span>}
                    </div>
                  </div>
                  <div className="flex gap-2">
                    <Button
                      size="md"
                      variant="surface"
                      color="neutral"
                      onClick={() => onEditPort(port)}
                      type="button"
                      data-testid="edit-port"
                    >
                      <Icon iconName="gear" />
                    </Button>
                    <Button
                      size="md"
                      variant="surface"
                      color="neutral"
                      onClick={() => onRemovePort(port)}
                      type="button"
                      data-testid="remove-port"
                    >
                      <Icon iconName="trash" />
                    </Button>
                  </div>
                </div>
              )
            })
          ) : (
            <div className="my-4 px-10 py-5 text-center">
              <Icon iconName="wave-pulse" className="text-neutral-300" />
              <p className="mb-3 mt-1 text-xs font-medium text-neutral-350">No port are set.</p>
            </div>
          )}
        </BlockContent>
        {children}
      </form>
    </Section>
  )
}

export default NetworkingSetting
