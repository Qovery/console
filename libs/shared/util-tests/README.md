# shared-util-tests

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test shared-util-tests` to execute the unit tests via [Jest](https://jestjs.io).
