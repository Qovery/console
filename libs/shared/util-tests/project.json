{
  "name": "shared-util-tests",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util-tests/src",
  "projectType": "library",
  "tags": ["scope:shared", "type:util"],
  "implicitDependencies": ["!shared-ui"],
  "targets": {
    "test": {
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "coverage": true
        }
      }
    }
  }
}
