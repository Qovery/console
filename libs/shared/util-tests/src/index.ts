export * from './lib/shared-util-tests'
