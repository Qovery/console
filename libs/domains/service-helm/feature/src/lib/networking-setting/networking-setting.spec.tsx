import selectEvent from 'react-select-event'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import { NetworkingSetting } from './networking-setting'

jest.mock('../hooks/use-kubernetes-services/use-kubernetes-services', () => ({
  ...jest.requireActual('../hooks/use-kubernetes-services/use-kubernetes-services'),
  useKubernetesServices: () => ({
    data: [],
    isLoading: false,
  }),
}))

describe('NetworkingSetting', () => {
  it('should match snapshot in empty state', () => {
    const { baseElement } = renderWithProviders(<NetworkingSetting ports={[]} onUpdatePorts={jest.fn()} />)
    expect(baseElement).toMatchSnapshot()
  })
  it('should match snapshot with ports', () => {
    const { baseElement } = renderWithProviders(
      <NetworkingSetting
        ports={[
          {
            internal_port: 1234,
            external_port: 443,
            service_name: 'My service',
            namespace: 'My namespace',
            protocol: 'HTTP',
            name: 'My service-p1234',
          },
          {
            internal_port: 4321,
            external_port: 443,
            service_name: 'My service 2',
            namespace: '',
            protocol: 'GRPC',
            name: 'My service 2-p4321',
          },
        ]}
        onUpdatePorts={jest.fn()}
      />
    )
    expect(baseElement).toMatchSnapshot()
  })
  it('should add port', async () => {
    const onUpdatePorts = jest.fn()
    const { userEvent } = renderWithProviders(<NetworkingSetting ports={[]} onUpdatePorts={onUpdatePorts} />)

    await userEvent.click(screen.getByText(/add port/i))

    screen.getByText('Set port')
    await userEvent.type(screen.getByLabelText(/service name/i), 'foo')
    await userEvent.type(screen.getByLabelText(/namespace/i), 'bar')
    await userEvent.type(screen.getByLabelText(/service port/i), '1234')
    await selectEvent.select(screen.getByLabelText(/protocol/i), ['GRPC'], {
      container: document.body,
    })
    await userEvent.click(screen.getByText(/create/i))

    expect(screen.queryByText('Set port')).not.toBeInTheDocument()

    expect(onUpdatePorts).toHaveBeenCalledWith([
      {
        service_name: 'foo',
        namespace: 'bar',
        internal_port: 1234,
        external_port: 443,
        protocol: 'GRPC',
        name: 'p1234-foo',
      },
    ])
  })
  it('should edit port', async () => {
    const onUpdatePorts = jest.fn()
    const { userEvent } = renderWithProviders(
      <NetworkingSetting
        ports={[
          {
            internal_port: 1234,
            external_port: 443,
            service_name: 'My service',
            namespace: 'My namespace',
            protocol: 'HTTP',
            name: 'My service-p1234',
          },
          {
            internal_port: 4321,
            external_port: 443,
            service_name: 'My service 2',
            namespace: '',
            protocol: 'GRPC',
            name: 'My service 2-p4321',
          },
        ]}
        onUpdatePorts={onUpdatePorts}
      />
    )

    await userEvent.click(screen.getAllByTestId('edit-port')[1])

    screen.getByText('Edit port')
    const serviceNameInput = screen.getByLabelText(/service name/i)
    await userEvent.clear(serviceNameInput)
    await userEvent.type(serviceNameInput, 'foo')
    const namespaceInput = screen.getByLabelText(/namespace/i)
    await userEvent.clear(namespaceInput)
    await userEvent.type(namespaceInput, 'bar')
    const portInput = screen.getByLabelText(/service port/i)
    await userEvent.clear(portInput)
    await userEvent.type(portInput, '1234')
    await selectEvent.select(screen.getByLabelText(/protocol/i), ['GRPC'], {
      container: document.body,
    })
    await userEvent.click(screen.getByText(/confirm/i))

    expect(screen.queryByText('Edit port')).not.toBeInTheDocument()

    expect(onUpdatePorts).toHaveBeenCalledWith([
      {
        internal_port: 1234,
        external_port: 443,
        service_name: 'My service',
        namespace: 'My namespace',
        protocol: 'HTTP',
        name: 'My service-p1234',
      },
      {
        service_name: 'foo',
        namespace: 'bar',
        internal_port: 1234,
        external_port: 443,
        protocol: 'GRPC',
        name: 'p1234-foo',
      },
    ])
  })

  it('should remove port', async () => {
    const onUpdatePorts = jest.fn()
    const { userEvent } = renderWithProviders(
      <NetworkingSetting
        ports={[
          {
            internal_port: 1234,
            external_port: 443,
            service_name: 'My service',
            namespace: 'My namespace',
            protocol: 'HTTP',
            name: 'My service-p1234',
          },
        ]}
        onUpdatePorts={onUpdatePorts}
      />
    )

    await userEvent.click(screen.getAllByTestId('remove-port')[0])

    await userEvent.type(screen.getByRole('textbox'), 'delete')
    await userEvent.click(screen.getByRole('button', { name: /confirm/i }))

    expect(onUpdatePorts).toHaveBeenCalledWith([])
  })
})
