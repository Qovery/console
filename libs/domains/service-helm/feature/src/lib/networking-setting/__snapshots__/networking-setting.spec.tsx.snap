// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkingSetting should match snapshot in empty state 1`] = `
<body>
  <div>
    <section
      class="flex flex-col items-start"
    >
      <div
        class="flex w-full flex-row justify-between"
      >
        <div>
          <h1
            class="text-neutral-400 text-xl font-bold mb-2"
          >
            Networking
          </h1>
          <p
            class="mb-2 text-sm text-neutral-350"
          >
            You can expose publicly over HTTP/gRPC the Kubernetes services deployed.
          </p>
          <span
            aria-controls="radix-:r6:"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="mb-5 cursor-pointer text-sm font-medium text-brand-500 transition hover:text-brand-600"
            data-state="closed"
            type="button"
          >
            How it works 
            <i
              aria-hidden="true"
              class="fa-regular fa-circle-question text-xs"
            />
          </span>
        </div>
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100 gap-2"
        >
          Add port
          <i
            aria-hidden="true"
            class="fa-solid fa-plus-circle "
          />
        </button>
      </div>
      <form
        class="w-full"
      >
        <div
          class="mb-5 rounded border border-neutral-250 bg-neutral-100 w-full"
          data-testid="block-content"
        >
          <div
            class="flex h-9 items-center justify-between border-b border-neutral-250 px-4"
          >
            <h2
              class="text-ssm font-medium text-neutral-400"
            >
              Services exposed publicly
            </h2>
          </div>
          <div
            class="p-0"
          >
            <div
              class="my-4 px-10 py-5 text-center"
            >
              <i
                aria-hidden="true"
                class="fa-solid fa-wave-pulse text-neutral-300"
              />
              <p
                class="mb-3 mt-1 text-xs font-medium text-neutral-350"
              >
                No port are set.
              </p>
            </div>
          </div>
        </div>
      </form>
    </section>
  </div>
</body>
`;

exports[`NetworkingSetting should match snapshot with ports 1`] = `
<body>
  <div>
    <section
      class="flex flex-col items-start"
    >
      <div
        class="flex w-full flex-row justify-between"
      >
        <div>
          <h1
            class="text-neutral-400 text-xl font-bold mb-2"
          >
            Networking
          </h1>
          <p
            class="mb-2 text-sm text-neutral-350"
          >
            You can expose publicly over HTTP/gRPC the Kubernetes services deployed.
          </p>
          <span
            aria-controls="radix-:rd:"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="mb-5 cursor-pointer text-sm font-medium text-brand-500 transition hover:text-brand-600"
            data-state="closed"
            type="button"
          >
            How it works 
            <i
              aria-hidden="true"
              class="fa-regular fa-circle-question text-xs"
            />
          </span>
        </div>
        <button
          class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100 gap-2"
        >
          Add port
          <i
            aria-hidden="true"
            class="fa-solid fa-plus-circle "
          />
        </button>
      </div>
      <form
        class="w-full"
      >
        <div
          class="mb-5 rounded border border-neutral-250 bg-neutral-100 w-full"
          data-testid="block-content"
        >
          <div
            class="flex h-9 items-center justify-between border-b border-neutral-250 px-4"
          >
            <h2
              class="text-ssm font-medium text-neutral-400"
            >
              Services exposed publicly
            </h2>
          </div>
          <div
            class="p-0"
          >
            <div
              class="flex w-full items-center justify-between gap-3 border-b border-neutral-250 px-5 py-4 last:border-0"
            >
              <div
                class="flex flex-col"
              >
                <span
                  class="text-sm font-medium text-neutral-400"
                >
                  My service
                </span>
                <div
                  class="flex flex-row gap-2 text-xs text-neutral-350"
                >
                  <span>
                    Service port: 
                    1234
                  </span>
                  <span>
                    Protocol: 
                    HTTP
                  </span>
                  <span>
                    Namespace: 
                    My namespace
                  </span>
                </div>
              </div>
              <div
                class="flex gap-2"
              >
                <button
                  class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-brand-500 text-sm h-9 px-3 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none bg-neutral-100 dark:bg-neutral-500 hover:[&:not(:active):not(:disabled)]:bg-white dark:hover:[&:not(:active):not(:disabled)]:bg-neutral-400"
                  data-testid="edit-port"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa-solid fa-gear "
                  />
                </button>
                <button
                  class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-brand-500 text-sm h-9 px-3 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none bg-neutral-100 dark:bg-neutral-500 hover:[&:not(:active):not(:disabled)]:bg-white dark:hover:[&:not(:active):not(:disabled)]:bg-neutral-400"
                  data-testid="remove-port"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa-solid fa-trash "
                  />
                </button>
              </div>
            </div>
            <div
              class="flex w-full items-center justify-between gap-3 border-b border-neutral-250 px-5 py-4 last:border-0"
            >
              <div
                class="flex flex-col"
              >
                <span
                  class="text-sm font-medium text-neutral-400"
                >
                  My service 2
                </span>
                <div
                  class="flex flex-row gap-2 text-xs text-neutral-350"
                >
                  <span>
                    Service port: 
                    4321
                  </span>
                  <span>
                    Protocol: 
                    GRPC
                  </span>
                </div>
              </div>
              <div
                class="flex gap-2"
              >
                <button
                  class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-brand-500 text-sm h-9 px-3 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none bg-neutral-100 dark:bg-neutral-500 hover:[&:not(:active):not(:disabled)]:bg-white dark:hover:[&:not(:active):not(:disabled)]:bg-neutral-400"
                  data-testid="edit-port"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa-solid fa-gear "
                  />
                </button>
                <button
                  class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-brand-500 text-sm h-9 px-3 rounded border border-neutral-250 dark:border-neutral-400 text-neutral-400 dark:text-white hover:[&:not(:active)]:border-neutral-300 dark:hover:[&:not(:active):not(:disabled)]:border-neutral-350 active:bg-neutral-150 disabled:text-neutral-300 disabled:bg-neutral-100 disabled:border-none bg-neutral-100 dark:bg-neutral-500 hover:[&:not(:active):not(:disabled)]:bg-white dark:hover:[&:not(:active):not(:disabled)]:bg-neutral-400"
                  data-testid="remove-port"
                  type="button"
                >
                  <i
                    aria-hidden="true"
                    class="fa-solid fa-trash "
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </form>
    </section>
  </div>
</body>
`;
